### PR TITLE
[MOD-6738] IndexComputer

### PR DESCRIPTION
**Pull request by meiravgri**
In this PR introduces two new components of the VecSimIndexAbstract: Preprocessor container and a calculator.

## Preprocessor container
The `PreprocessorsContainerAbstract` API supports processing blobs for storage, for query (graph search) or both.
It is also responsible to copy the original blob if it needs to be processed and wrapping it with the appropriate deleted.
`PreprocessorsContainerAbstract` preprocessing includes alignment of a query blob.

### Multi preprocessors
`MultiPreprocessorsContainer` extends `PreprocessorsContainerAbstract` by holding an array of pointers to `PreprocessorInterface` objects, each responsible for a different processing step.
Currently, we only have `CosinePreprocessor`, which normalizes vectors if the index type is Cosine. Calling CosinePreprocessor::preprocess with a blob pointing to the same memory address will result in one normalization call and a the returned blobs will point to the same memory.

***NOTE***: in tiered index, we assume that the vectors are processed before being inserted into the backend index, so the frontend index will be of type `VecSimMetric_Cosine`, but internally doesn't hold a cosine preprocessor.

The processed blobs have a scope lifetime and will be released automatically. It is assumed that they are copied if their lifetime needs to be extended (for storage purposes for example).

## Distance Calculator
The distance calculator is defined according to the distance function signature.
It holds the distance function of the abstract index.
The distance calculation API of all Distance Calculator classes is: calc_dist(v1,v2,dim), but internally they will call the distance function according to the template signature.

## Index API changes

- An index of type `VecSimIndexAbstract` is responsible for preprocessing a blob before performing any operation.
- It includes adding a new vector and processing a query before searching in the index.
- All `*Wrapper` functions were removed.
- As for the tiered index, it is assumed that the backend index receives a blob that was preprocessed by the frontend index. The backend index can perform additional preprocessing if needed.

### CODE REVIEW

File: src/VecSim/spaces/computer/computer.h [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    }
+    return -1;
+}
+
+template <typename DistType, typename DistFuncType>
+ProcessedBlobs
+IndexComputerExtended<DistType, DistFuncType>::preprocess(const void *original_blob,
+                                                          size_t processed_bytes_count) const {
+    // No preprocessors were added yet.
+    if (preprocessors[0] == nullptr) {
```
Comment (GuyAv46) [inline src/VecSim/spaces/computer/computer.h:246]: Do we need a dynamic mechanism for adding preprocessors in runtime?

File: src/VecSim/spaces/computer/computer.h [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+            preprocessors[i] = preprocessor;
+            return i + 1 >= n_preprocessors ? 0 : i + 1;
+        }
+    }
+    return -1;
+}
+
+template <typename DistType, typename DistFuncType>
+ProcessedBlobs
+IndexComputerExtended<DistType, DistFuncType>::preprocess(const void *original_blob,
```
Comment (GuyAv46) [inline src/VecSim/spaces/computer/computer.h:243]: I don't get why this function is so complex.
Why isn't it a single loop over the preprocessors, and calling it on one or both blobs depends on its flags?
Why do we need to support an incomplete state where the preprocessors have not been initialized yet?

File: src/VecSim/spaces/computer/preprocessor.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    virtual void preprocess(const void *original_blob, MemoryUtils::unique_blob &storage_blob,
+                            MemoryUtils::unique_blob &query_blob,
+                            PreprocessParams &params) const = 0;
+    virtual void preprocessForStorage(MemoryUtils::unique_blob &blob) const = 0;
+    virtual void preprocessQuery(MemoryUtils::unique_blob &blob) const = 0;
+
+    virtual bool hasQueryPreprocessor() const = 0;
+    virtual bool hasStoragePreprocessor() const = 0;
+
+    struct PreprocessParams {
```
Comment (GuyAv46) [inline src/VecSim/spaces/computer/preprocessor.h:27]: Is this the state of the blobs? I'm not sure it should belong to the preprocessor. I'm also not sure why we need to keep it dynamic - aren't we supposed to create the right pipeline and then just use it without validations?

File: src/VecSim/vec_sim_tiered_index.h [hunk 4]
```diff
@@ -158,12 +141,13 @@ VecSimTieredIndex<DataType, DistType>::topKQuery(const void *queryBlob, size_t k
 
         // Simply query the main index and return the results while holding the lock.
         this->mainIndexGuard.lock_shared();
-        auto res = this->backendIndex->topKQuery(queryBlob, k, queryParams);
+        auto res = this->backendIndex->topKQueryWrapper(queryBlob, k, queryParams);
```
Comment (GuyAv46) [inline src/VecSim/vec_sim_tiered_index.h:144]: if we reached the `topKQuery` function of `VecSimTieredIndex`, doesn't it mean we already processed the blob when `topKQueryWrapper` was called?

File: src/VecSim/vec_sim_tiered_index.h [hunk 5]
```diff
@@ -175,7 +159,7 @@ VecSimTieredIndex<DataType, DistType>::topKQuery(const void *queryBlob, size_t k
 
         // Lock the main index and query it.
         this->mainIndexGuard.lock_shared();
-        auto main_results = this->backendIndex->topKQuery(queryBlob, k, queryParams);
+        auto main_results = this->backendIndex->topKQueryWrapper(queryBlob, k, queryParams);
```
Comment (GuyAv46) [inline src/VecSim/vec_sim_tiered_index.h:162]: same here

File: src/VecSim/vec_sim_tiered_index.h [hunk 6]
```diff
@@ -210,7 +194,7 @@ VecSimTieredIndex<DataType, DistType>::rangeQuery(const void *queryBlob, double
 
         // Simply query the main index and return the results while holding the lock.
         this->mainIndexGuard.lock_shared();
-        auto res = this->backendIndex->rangeQuery(queryBlob, radius, queryParams);
+        auto res = this->backendIndex->rangeQueryWrapper(queryBlob, radius, queryParams);
```
Comment (GuyAv46) [inline src/VecSim/vec_sim_tiered_index.h:197]: same here

File: src/VecSim/vec_sim_tiered_index.h [hunk 7]
```diff
@@ -230,7 +215,7 @@ VecSimTieredIndex<DataType, DistType>::rangeQuery(const void *queryBlob, double
 
         // Lock the main index and query it.
         this->mainIndexGuard.lock_shared();
-        auto main_results = this->backendIndex->rangeQuery(queryBlob, radius, queryParams);
+        auto main_results = this->backendIndex->rangeQueryWrapper(queryBlob, radius, queryParams);
```
Comment (GuyAv46) [inline src/VecSim/vec_sim_tiered_index.h:218]: same here

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 8]
```diff
@@ -243,7 +246,7 @@ BruteForceIndex<DataType, DistType>::topKQuery(const void *queryBlob, size_t k,
             delete TopCandidates;
             return rep;
         }
-        auto score = this->distFunc(vector, queryBlob, this->dim);
+        auto score = this->indexComputer->calcDistance(vector, queryBlob, this->dim);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.h:249]: Possible, i added `calcDistance` later for external usage

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 9]
```diff
@@ -954,7 +961,7 @@ VecSimQueryReply *TieredHNSWIndex<DataType, DistType>::TieredHNSW_BatchIterator:
         // it until the iterator is depleted or freed.
         this->index->mainIndexGuard.lock_shared();
         this->hnsw_iterator =
-            this->index->backendIndex->newBatchIterator(getQueryBlob(), queryParams);
+            this->index->backendIndex->newBatchIteratorWrapper(getQueryBlob(), queryParams);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:964]: newBatchIteratorWrapper processes the blob before sending it to newBatchIterator

File: src/VecSim/algorithms/hnsw/hnsw_single.h [hunk 10]
```diff
@@ -140,22 +143,17 @@ int HNSWIndex_Single<DataType, DistType>::deleteVector(const labelType label) {
...
     bool label_exists = false;
-    // Note that is it the caller responsibility to ensure that this label doesn't exist in the
-    // index and increase the element count before calling this, if auxiliaryCtx is *not* NULL.
-    if (auxiliaryCtx == nullptr) {
-        if (labelLookup.find(label) != labelLookup.end()) {
-            label_exists = true;
-            // Remove the vector in place if override allowed (in non-async scenario).
-            deleteVector(label);
-        }
+    if (labelLookup.find(label) != labelLookup.end()) {
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_single.h:149]: why not

File: src/VecSim/index_factories/brute_force_factory.cpp [hunk 11]
```diff
@@ -43,30 +44,49 @@ static AbstractIndexInitParams NewAbstractInitParams(const VecSimParams *params)
     return abstractInitParams;
 }
 
-VecSimIndex *NewIndex(const VecSimParams *params) {
+VecSimIndex *NewIndex(const VecSimParams *params, bool is_normalized) {
     const BFParams *bfParams = &params->algoParams.bfParams;
     AbstractIndexInitParams abstractInitParams = NewAbstractInitParams(params);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/index_factories/brute_force_factory.cpp:49]: no, because no index type sends normalized blobs to the BF index.
Not sure that we have a test suit for our factory so I can imitate such case 🤔

File: src/VecSim/index_factories/factory_utils.h [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+                                                     VecSimMetric metric, size_t dim) {
+    unsigned char alignment = 0;
+    spaces::dist_func_t<DistType> distFunc =
+        spaces::GetDistFunc<DataType, DistType>(metric, dim, &alignment);
+    DistanceCalculatorCommon<DistType> *distance_calculator =
+        new (allocator) DistanceCalculatorCommon<DistType>(allocator, distFunc);
+
+    if (metric == VecSimMetric_Cosine) {
+        IndexComputerExtended<DistType, spaces::dist_func_t<DistType>> *indexComputer =
+            new (allocator) IndexComputerExtended<DistType, spaces::dist_func_t<DistType>>(
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/index_factories/factory_utils.h:25]: I prefer not to use default values. Help the compiler help and let it alert when someone misuses the API.

File: src/VecSim/index_factories/factory_utils.h [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+IndexComputerAbstract<DistType> *CreateIndexComputer(std::shared_ptr<VecSimAllocator> allocator,
+                                                     VecSimMetric metric, size_t dim) {
+    unsigned char alignment = 0;
+    spaces::dist_func_t<DistType> distFunc =
+        spaces::GetDistFunc<DataType, DistType>(metric, dim, &alignment);
+    DistanceCalculatorCommon<DistType> *distance_calculator =
+        new (allocator) DistanceCalculatorCommon<DistType>(allocator, distFunc);
+
+    if (metric == VecSimMetric_Cosine) {
+        IndexComputerExtended<DistType, spaces::dist_func_t<DistType>> *indexComputer =
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/index_factories/factory_utils.h:24]: I wrote the type explicitly to make sure everything is implemented as I intended 😅 I can confidently convert now!

File: src/VecSim/memory/vecsim_malloc.cpp [hunk 14]
```diff
@@ -45,6 +45,11 @@ void *VecSimAllocator::allocate_aligned(size_t size, unsigned char alignment) {
     if (!alignment) {
         return allocate(size);
     }
+
+    return allocate_force_aligned(size, alignment);
+}
+
+void *VecSimAllocator::allocate_force_aligned(size_t size, unsigned char alignment) {
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/memory/vecsim_malloc.cpp:52]: The cost of the wrapper function is 0 so why not 🤷 
ill add an assertion

File: src/VecSim/spaces/computer/calculator.h [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+template <typename DistType>
+class DistanceCalculatorCommon
+    : public DistanceCalculatorAbstract<DistType, spaces::dist_func_t<DistType>> {
+public:
+    DistanceCalculatorCommon(std::shared_ptr<VecSimAllocator> allocator,
+                             spaces::dist_func_t<DistType> dist_func)
+        : DistanceCalculatorAbstract<DistType, spaces::dist_func_t<DistType>>(allocator,
+                                                                              dist_func) {}
+
+    virtual DistType calcDistance(const void *v1, const void *v2, size_t dim) const {
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/calculator.h:33]: To allow the index to hold DistanceCalculatorAbstract * but call the underlying instance implementation

File: src/VecSim/spaces/computer/computer.h [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+
+    // Delete copy constructor and assignment operator to avoid copying unique_ptr
+    ProcessedBlobs(const ProcessedBlobs &) = delete;
+    ProcessedBlobs &operator=(const ProcessedBlobs &) = delete;
+
+    const void *getStorageBlob() const { return storage_blob.get(); }
+    const void *getQueryBlob() const { return query_blob.get(); }
+
+    static MemoryUtils::unique_blob uniqueWithDummyDeleter(void *ptr) {
+        return MemoryUtils::unique_blob(ptr, MemoryUtils::dummyFreeAllocation);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/computer.h:42]: the return type should be uniform...

File: src/VecSim/spaces/computer/computer.h [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+                                                       size_t processed_bytes_count) const {
+    auto storage_blob = preprocessForStorage(original_blob, processed_bytes_count);
+    auto query_blob = preprocessQuery(original_blob, processed_bytes_count);
+    return ProcessedBlobs(std::move(storage_blob), std::move(query_blob));
+}
+
+template <typename DistType, typename DistFuncType>
+MemoryUtils::unique_blob IndexComputerBasic<DistType, DistFuncType>::preprocessForStorage(
+    const void *original_blob, size_t processed_bytes_count) const {
+    return MemoryUtils::unique_blob(const_cast<void *>(original_blob),
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/computer.h:189]: IndexComputerBasic doesn't hold preprocessors, it only aligns query blobs
so storage blobs are not aligned, but return wrapped in a unique ptr
The cast is for uniformity in the return value

File: src/VecSim/spaces/computer/computer.h (hunk 1 above)
Comment (meiravgri) [reply src/VecSim/spaces/computer/computer.h:246]: Yes for scalar quantization, where we build the preprocessor after the training phase

File: src/VecSim/spaces/computer/preprocessor.h (hunk 3 above)
Comment (meiravgri) [reply src/VecSim/spaces/computer/preprocessor.h:27]: The computer flags should determine type of unique ptr (frees the blob at the end or not) and the preprocessor params hint the preprocessor if the unique pte is already populated or needs to be initialized.

File: src/VecSim/vec_sim_tiered_index.h (hunk 4 above)
Comment (meiravgri) [reply src/VecSim/vec_sim_tiered_index.h:144]: The frontend computer processed it. The backend computer might have additional processing to do

File: src/VecSim/spaces/computer/computer.h [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+    // Allocate and copy the blob only if the original blob is not aligned.
+    MemoryUtils::unique_blob maybeCopyToAlignedMem(const void *original_blob,
+                                                   size_t blob_bytes_count) const;
+};
+
+template <typename DistType, typename DistFuncType>
+class IndexComputerExtended : public IndexComputerBasic<DistType, DistFuncType> {
+protected:
+    PreprocessorAbstract **preprocessors;
+    const size_t n_preprocessors;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/computer.h:113]: Vector as an overkill since we know at compile time the size of the allocation
The only alternative is to use `n_preprocessors` as a template argument

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:964]: But why do we have to process it again?

[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:964]: Because the frontend and the backend index might perform different processing

[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:964]: Currently, we don't have such scenario so the blob will be processed only once.

File: src/VecSim/memory/vecsim_malloc.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/memory/vecsim_malloc.cpp:53]: IMO it complicates the code base for no reason.
`allocate_aligned` was implemented to get an aligned allocation, and is meant to be used when we want an aligned allocation. It supports the `alignment=0` edge case as a feature so we don't need to verify we have an alignment requirement.

File: src/VecSim/spaces/computer/calculator.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/spaces/computer/calculator.h:33]: We need the `DistanceCalculatorAbstract` to have this virtual signature but here we only need `override` no?

[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/calculator.h:33]: Yes you don't have to redeclare it as virtual explicitly, but I think it improves readability by emphasizing the source of this function

File: src/VecSim/memory/vecsim_malloc.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/memory/vecsim_malloc.cpp:53]: We check that the alignment is not 0 in the computer, so no point in rechecking it in `allocate_aligned`

File: src/VecSim/spaces/computer/calculator.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/calculator.h:33]: i can replace it with override you are right.
Sometimes the compiler complains when override is missing, i wonder why it did;t this time 🤔

File: src/VecSim/memory/vecsim_malloc.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/memory/vecsim_malloc.cpp:53]: will revert

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:964]: newBatchIteratorWrapper will replace newBatchIterator

**Review by GuyAv46 — APPROVED** [review]

File: src/VecSim/spaces/computer/preprocessor.h [hunk 19]
```diff
@@ -14,21 +14,14 @@ class PreprocessorAbstract : public VecsimBaseObject {
...
-    virtual bool hasQueryPreprocessor() const = 0;
-    virtual bool hasStoragePreprocessor() const = 0;
-
-    struct PreprocessParams {
-        const size_t processed_bytes_count;
-        bool is_populated_storage;
-        bool is_populated_query;
-    };
+    virtual void preprocess(const void *original_blob, void *&storage_blob, void *&query_blob,
+                            size_t processed_bytes_count, unsigned char alignment) const = 0;
```
Comment (GuyAv46) [inline src/VecSim/spaces/computer/preprocessor.h:18]: shouldn't the `processed_bytes_count` be a reference in case it is changed by the preprocessor?

File: src/VecSim/spaces/computer/computer.h [hunk 20]
```diff
@@ -197,161 +184,125 @@ IndexComputerBasic<DistType, DistFuncType>::preprocessQuery(const void *original
...
-template <typename DistType, typename DistFuncType>
-MemoryUtils::unique_blob IndexComputerExtended<DistType, DistFuncType>::preprocessForStorage(
-    const void *original_blob, size_t processed_bytes_count) const {
-
-    // Not a storage preprocessor or no preprocessors were added yet.
-    if (!(this->flags & Flags::STORAGE) || preprocessors[0] == nullptr) {
-        return IndexComputerBasic<DistType, DistFuncType>::preprocessForStorage(
-            original_blob, processed_bytes_count);
+    // If they point to the same memory, we need to free only one of them.
+    if (storage_blob == query_blob) {
```
Comment (GuyAv46) [inline src/VecSim/spaces/computer/computer.h:243]: we also assume at least one is not null right? consider commenting on this

**Review by GuyAv46 — APPROVED** [review]

File: src/VecSim/spaces/computer/preprocessor.h (hunk 19 above)
Comment (meiravgri) [reply src/VecSim/spaces/computer/preprocessor.h:18]: i havn't yet decicded how we would handle a preprocessor that changes the allocation size 
for now we assume uniform size

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 21]
```diff
@@ -707,10 +711,12 @@ int TieredHNSWIndex<DataType, DistType>::addVector(const void *blob, labelType l
         if (!this->backendIndex->isMultiValue()) {
             ret -= this->deleteVector(label);
         }
+
+        auto storage_blob = this->frontendIndex->processForStorage(blob);
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:715]: Let's explain it in a comment

File: tests/unit/test_hnsw_tiered.cpp [hunk 22]
```diff
@@ -58,7 +63,7 @@ TYPED_TEST(HNSWTieredIndexTest, CreateIndexInstance) {
     // Create TieredHNSW index instance with a mock queue.
     HNSWParams params = {.type = TypeParam::get_index_type(),
                          .dim = 4,
-                         .metric = VecSimMetric_L2,
+                         .metric = VecSimMetric_IP,
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:66]: Why the change?

File: tests/unit/test_hnsw_tiered.cpp [hunk 23]
```diff
@@ -3090,6 +3095,192 @@ TYPED_TEST(HNSWTieredIndexTest, bufferLimitAsync) {
     EXPECT_EQ(tiered_index->indexLabelCount(), n_labels);
 }
 
+TYPED_TEST(HNSWTieredIndexTestBasic, RangeSearchCosine) {
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:3098]: What is the purpose of these 3 tests? validating that cosine preprocessor is working as expected?

File: tests/unit/test_hnsw_tiered.cpp [hunk 24]
```diff
@@ -3800,3 +3991,297 @@ TYPED_TEST(HNSWTieredIndexTestBasic, switchDeleteModes) {
...
+            blob = this->allocator->allocate_aligned(processed_bytes_count, alignment);
+            memcpy(blob, original_blob, processed_bytes_count);
+        }
+        for (size_t i = 0; i < dim; i++) {
+            static_cast<DataType *>(blob)[i] *= 2;
+        }
+    }
+};
+
+TYPED_TEST(HNSWTieredIndexTestBasic, indexComputerHNSWPreprocessor) {
```
Comment (alonre24) [inline tests/unit/test_hnsw_tiered.cpp:4169]: Nice test

**Review by alonre24 — COMMENTED** [review]
Great work! 
Some of the comments were the basis for our discussion, others are more suggestions/questions

File: src/VecSim/memory/vecsim_malloc.cpp [hunk 25]
```diff
@@ -9,6 +9,7 @@
 #include <memory>
 #include <string.h>
 #include <sys/param.h>
+#include <cassert>
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/memory/vecsim_malloc.cpp:12]: i guess yes

File: src/VecSim/spaces/computer/computer.h [hunk 26]
```diff
@@ -0,0 +1,309 @@
...
+                break;
+
+            delete pp;
+        }
+    }
+
+    /** @returns On success, next uninitialized index, or 0 in case capacity is reached (after
+     * inserting the preprocessor). -1 if capacity is full and we failed to add the preprocessor.
+     */
+    int addPreprocessor(PreprocessorAbstract *preprocessor);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/computer.h:144]: We always know at compilation time the size of the array so we can save the memory overhead of a dynamic array.
To avoid checks we can save the current number of initiazlized preprocessor, but assuming we are not expecting to have a large amount of preprocessors, i think it won't make any diffrence.

File: src/VecSim/spaces/computer/computer.h [hunk 27]
```diff
@@ -0,0 +1,309 @@
...
+
+/* ======================= IndexComputerExtended ======================= */
+
+template <typename DistType, typename DistFuncType, size_t n_preprocessors>
+int IndexComputerExtended<DistType, DistFuncType, n_preprocessors>::addPreprocessor(
+    PreprocessorAbstract *preprocessor) {
+    for (size_t i = 0; i < n_preprocessors; i++) {
+        if (preprocessors[i] == nullptr) {
+            preprocessors[i] = preprocessor;
+            return i + 1 >= n_preprocessors ? 0 : i + 1;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/computer.h:215]: There's a description in the function's declartion.
Also changed the variables names to make it clearer

File: src/VecSim/vec_sim_index.h [hunk 28]
```diff
@@ -204,44 +223,32 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
...
+MemoryUtils::unique_blob
+VecSimIndexAbstract<DataType, DistType>::processQuery(const void *queryBlob) const {
+    return this->indexComputer->preprocessQuery(queryBlob, this->dataSize);
+}
+
+template <typename DataType, typename DistType>
+MemoryUtils::unique_blob
+VecSimIndexAbstract<DataType, DistType>::processForStorage(const void *original_blob) const {
+    return this->indexComputer->preprocessForStorage(original_blob, this->dataSize);
+}
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/vec_sim_index.h:254]: As discussed, the IndexComputer is a private member of the index and should not be exposed to other objects.
Instead, the index should provide preprocessing functions as part of its API.

File: src/VecSim/spaces/computer/preprocessor.h [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+    void preprocessQuery(const void *original_blob, void *&blob, size_t processed_bytes_count,
+                         unsigned char alignment) const override {
+        if (blob == nullptr) {
+            blob = this->allocator->allocate_aligned(processed_bytes_count, alignment);
+            memcpy(blob, original_blob, processed_bytes_count);
+        }
+        normalize_func(blob, this->dim);
+    }
+
+    void preprocessQueryInPlace(void *blob, size_t processed_bytes_count,
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/preprocessor.h:85]: Currently yes since we already hold a copy of the original blob in the batch iterator object.

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 30]
```diff
@@ -197,7 +197,7 @@ class TieredHNSWIndex : public VecSimTieredIndex<DataType, DistType> {
     VecSimInfoIterator *infoIterator() const override;
     VecSimBatchIterator *newBatchIterator(const void *queryBlob,
                                           VecSimQueryParams *queryParams) const override {
-        size_t blobSize = this->backendIndex->getDim() * sizeof(DataType);
+        size_t blobSize = this->frontendIndex->getDim() * sizeof(DataType);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:200]: In tiered index it is always assumed that the frontend index gets the raw data

File: src/VecSim/index_factories/brute_force_factory.cpp [hunk 31]
```diff
@@ -43,30 +44,49 @@ static AbstractIndexInitParams NewAbstractInitParams(const VecSimParams *params)
...
 }
 
-VecSimIndex *NewIndex(const BFParams *bfparams, const AbstractIndexInitParams &abstractInitParams) {
+VecSimIndex *NewIndex(const BFParams *bfparams, const AbstractIndexInitParams &abstractInitParams,
+                      bool is_normalized) {
+    // If the index metric is Cosine, and is_normalized == true, we will skip normalizing vectors
+    // and query blobs.
+    VecSimMetric metric;
+    if (is_normalized && bfparams->metric == VecSimMetric_Cosine) {
+        metric = VecSimMetric_IP;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/index_factories/brute_force_factory.cpp:59]: Currently, no such scenario.
In the future it might be useful for a tired index where the backend index is BF (sq for example)

File: tests/unit/test_hnsw_tiered.cpp (hunk 24 above)
Comment (meiravgri) [reply tests/unit/test_hnsw_tiered.cpp:4169]: 😊

File: tests/unit/test_hnsw_tiered.cpp (hunk 23 above)
Comment (meiravgri) [reply tests/unit/test_hnsw_tiered.cpp:3098]: Yes

File: src/VecSim/index_factories/factory_utils.h [hunk 32]
```diff
@@ -0,0 +1,60 @@
+/*
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/index_factories/factory_utils.h:1]: I still think that we should have a separate file for the index computer factory (and maybe later on a file for the data container factory) - this is an important component that needs to be highlighted among the factory

File: src/VecSim/memory/memory_utils.h [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+ *the Server Side Public License v1 (SSPLv1).
+ */
+
+#pragma once
+#include <functional>
+
+namespace MemoryUtils {
+
+using alloc_deleter_t = std::function<void(void *)>;
+using unique_blob = std::unique_ptr<void, alloc_deleter_t>;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/memory/memory_utils.h:13]: no, the default deleter is a function calling delete on the pointer

File: src/VecSim/index_factories/factory_utils.h (hunk 32 above)
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/index_factories/factory_utils.h:1]: ok

**Review by alonre24 — DISMISSED** [review]
Let's go!!

**Review by alonre24 — APPROVED** [review]
